File: 0061. Rotate List M.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if (head == null || head.next == null || k == 0) return head;
        ListNode ptr = head;
        int cnt=1; 
        while (ptr.next!=null) {
            ptr = ptr.next;
            cnt++; 
        }
        ptr.next=head;
        ptr = head;
        int no = cnt - (k % cnt);
        for (int i=1; i<no; i++) {
            ptr=ptr.next;
        }
        head = ptr.next;
        ptr.next = null;
        return head;
    }
}

File: 0061. Rotate List M.cpp
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if (!head || !head->next || k == 0) return head;

        // Step 1: Count length
        ListNode* temp = head;
        int cnt = 1;
        while (temp->next) {
            cnt++;
            temp = temp->next;
        }

        // Step 2: Make circular
        temp->next = head;

        // Step 3: Normalize k
        k = k % cnt;
        if (k == 0) {
            temp->next = nullptr; // break circle
            return head;
        }

        // Step 4: Find new head
        int stepsToNewHead = cnt - k;
        temp = head;
        for (int i = 1; i < stepsToNewHead; i++) {
            temp = temp->next;
        }

        // Step 5: Break circle
        head = temp->next;
        temp->next = nullptr;

        return head;
    }
};